### PR TITLE
PERF: quantile now operates per block boosting perf / fix quantile with nan

### LINKED ISSUE #11623: Pandas quantile function very slow
The quantile function is almost 10 000 times slower than the equivalent percentile function in numpy. See code below:

```python
import time
import pandas as pd
import numpy as np

q = np.array([0.1,0.4,0.6,0.9])
data = np.random.randn(10000, 4)
df = pd.DataFrame(data, columns=['a', 'b', 'c', 'd'])
time1 = time.time()
pandas_quantiles = df.quantile(q, axis=1)
time2 = time.time()
print 'Pandas took %0.3f ms' % ((time2-time1)*1000.0)

time1 = time.time()
numpy_quantiles = np.percentile(data, q*100, axis=1)
time2 = time.time()
print 'Numpy took %0.3f ms' % ((time2-time1)*1000.0)

print (pandas_quantiles.values == numpy_quantiles).all()
# Output:
# Pandas took 15337.531 ms
# Numpy took 1.653 ms
# True

### LINKED ISSUE #13098: REGRP: Series.quantile returns NaN
#### Code Sample, a copy-pastable example if possible
```python
import pandas as pd
import numpy
s = pd.Series([1, 2, 3, 4, numpy.nan])
s.quantile(0.5)
nan
```

#### Expected Output
I would expect 2.5 as output (as with version 0.17.1).

#### output of ``pd.show_versions()``

commit: None
python: 2.7.6.final.0
python-bits: 64
OS: Linux
OS-release: 3.13.0-85-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: de_DE.UTF-8

pandas: 0.18.1
nose: None
pip: 1.5.4
setuptools: 2.2
Cython: None
numpy: 1.11.0
scipy: 0.16.1
statsmodels: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.5.3
pytz: 2016.4
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: 1.5.1
openpyxl: 2.3.2
xlrd: 0.9.4
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
boto: None
pandas_datareader: None

### PR DESCRIPTION
**Pull request by jreback**
closes #11623
closes #13098
There is a slight API change in that the returned Series are now named by the quantile.

this PR

```
In [1]: df = DataFrame(np.random.randn(1000, 3), columns=list('ABC'))

In [2]: %timeit df.quantile([0.1, 0.5], axis=1)
1000 loops, best of 3: 830 µs per loop
```

0.18.1

```
In [3]: df.quantile([0.1, 0.5], axis=1)
In [1]: df = DataFrame(np.random.randn(1000, 3), columns=list('ABC'))

In [2]: %timeit df.quantile([0.1, 0.5], axis=1)
1 loop, best of 3: 317 ms per loop
```

### CODE REVIEW

**Comment by jreback** [comment]
cc @sinhrks

**Comment by jreback** [comment]
@sinhrks this also resolves some thorny issues with ``.replace`` and datetime w/tz (now the conversion happens at a much lower level).

File: pandas/core/internals.py [hunk 1]
```diff
@@ -1260,32 +1267,117 @@ def equals(self, other):
             return False
         return array_equivalent(self.values, other.values)
 
-    def quantile(self, qs, mgr=None, **kwargs):
+    def quantile(self, qs, interpolation='linear', axis=0, mgr=None, **kwargs):
```
Comment (sinhrks) [inline pandas/core/internals.py]: ``kwargs`` is not used.

File: pandas/tests/frame/test_quantile.py [hunk 2]
```diff
@@ -256,12 +262,14 @@ def test_quantile_box(self):
                         'C': [pd.Timedelta('1 days'),
                               pd.Timedelta('2 days'),
                               pd.Timedelta('3 days')]})
+
         res = df.quantile(0.5, numeric_only=False)
+
         # when squeezed, result.name is explicitly reset
```
Comment (sinhrks) [inline pandas/tests/frame/test_quantile.py]: we can remove this comment now.

**Comment by sinhrks** [comment]
Thanks. I've at a loss how to achieve the behavior you did in ``reduction``. 

BTW, we don't seem to test ``.quantile(axis=1)`` with mixed dtype currently. It's nice to be added.

File: pandas/tests/frame/test_quantile.py (hunk 2 above)
Comment (jreback) [inline pandas/tests/frame/test_quantile.py]: done

File: pandas/core/internals.py (hunk 1 above)
Comment (jreback) [inline pandas/core/internals.py]: umm no.

**Comment by jreback** [comment]
ok added test for mixed, though not a whole lot you can do with it.

**Pull request closed without merging**